EIPS/eip-2935.md

```diff
@@ -12,90 +12,93 @@ created: 2020-09-03
 
 For resolving the `BLOCKHASH` opcode this fork onwards (`block.timestamp >= FORK_TIMESTAMP`), switch the logic to:
 
 ```python
 def resolve_blockhash(block: Block, state: State, arg: uint64):
   # check the wrap around range
-  if arg >= block.number or arg < max(block.number - MAX_HISTORY_SERVE_WINDOW, 0)
+  if arg >= block.number or arg < max(block.number - HISTORY_SERVE_WINDOW, 0)
     return 0
 
-  return state.load_slot(HISTORY_STORAGE_ADDRESS, arg)
+  return state.load_slot(HISTORY_STORAGE_ADDRESS, arg % HISTORY_SERVE_WINDOW)
 ```
 
-Note that the above logic allows access deeper than `MIN_HISTORY_SERVE_WINDOW` if it exists all the way upto `MAX_HISTORY_SERVE_WINDOW`.
-
-Edge cases:
+Some activation scenarios:
 
  * For the fork to be activated at genesis, no history is written to the genesis state, and at the start of block `1`, genesis hash will be written as a normal operation to slot `0`.
  * for activation at block `1`, only genesis hash will be written at slot `0` as there is no additional history that needs to be persisted.
  * for activation at block `32`, block `31`'s hash will be written to slot `31` and additonal history for `0..30`'s hashes will be persisted, so all in all `0..31`'s hashes.
- * for activation at block `1000`, block `744-999`'s hashes will be presisted in the slot and `BLOCKHASH` for `733` or less would resolve to `0` as only `MIN_HISTORY_SERVE_WINDOW` is persisted.
+ * for activation at block `1000`, block `744-999`'s hashes will be presisted in the slot and `BLOCKHASH` for `733` or less would resolve to `0` as only `HISTORY_SERVE_WINDOW` can be served.
 
-### [EIP-158](./eip-158.md) exception
+### [EIP-158](./eip-158.md) handling
 
-This address is currently exempt from [EIP-158](./eip-158.md) cleanup in Kaustinen Verkle Testnet but there are two ways this could be addressed before this EIP is adopted by ACD:
+This address is currently exempt from [EIP-158](./eip-158.md) cleanup in Kaustinen Verkle Testnet but we plan to address this in the following way:
 
-* Update the nonce to 1 in the fork block, or
-* Deploy a contract à la [EIP-4788](./eip-4788.md) with `BLOCKHASH` opcode delegating call to this contract with appropriate args.
-
-While the second option looks more elegant, it has a higher complexity as well as gas consumption considerations.
+* Deploy a contract à la [EIP-4788](./eip-4788.md) which just supports `get` method to resolve the BLOCKHASH as per the logic defined above (and use the generated address as the BLOCKHASH contract address).
+* While the clients are expected to directly read from state (or maintain and serve from memory) to resolve BLOCKHASH opcode, this contract's `get` could be invoked by transaction (via another contract or directly) leading to a normal contract execution (and gas consumption) as per the semantics of the contract call.
```

Comment:
This is fine, but as note: we will need a similar process like 4788 regarding auditing this contract etc etc in order to ship this :)

Reply:
With 4788 there are also checks in the contract (if timestamp does not match storage, then throw). We now seem to treat this contract on a different level: where is the "requested block hash >= block.number - 256" check? Is this in the contract, or is this done "outside" the contract? This is not clear from the current description (at least, for me)

Reply:
contract will implement `resolve_blockhash` logic so yes the check goes inside.

4788 worries about timestamp because could be missed slots, but no such thing as missed numbers (if i understand correctly)

Reply:
Yes to be clear: if one would call `resolve_blockhash(n)` and then calldata `n` would just return `sload(n % HISTORY_SERVE_WINDOW)` then we would be in trouble, but if there is this check in the contract that we are within range (i.e. we are within 256 blocks) then we are fine. I think I would just like an explicit note here such that the "shortcut" implementation in clients (i.e. not call the contract but write/read to storage immediately) has a consistent check in the EVM contract :smile:

Reply:
there is a check there, could you confirm if that is correct?

Reply:
If the check is also in the EVM bytecode then it looks correct!